### PR TITLE
Add reverse proxy endpoint.

### PR DESCRIPTION
**Pull request by xetorthio**
It works by using the `Host` of the request. When it receives something in
the form of: `<node>.<session>.play-with-docker.com` it does a reverse
proxy http request to `node`, validating that the `node` actually belongs
to the `session`.

If the node has a prefix `ip` and continues with a valid IP address
where the dots where replaces by underscores (like `ip10_0_0_1`) then it
will remove the `ip` prefix and and replace the underscores by dots, and
assume it is an ip address.

**There is an important assumption that the daemon container is called `pwd`**

### CODE REVIEW

**Comment by xetorthio** [comment]
:bell: @marcosnils @jpetazzo

**Comment by jpetazzo** [comment]
That looks very promising!

A few remarks:
- people will also want to specify a port number
- when I wrote my proxy recipe, I wondered whether to use the session ID in the URL or not, and eventually decided against it, because I wanted to allow people to share URLs without disclosing their session ID; but I guess it could go both ways

**Comment by xetorthio** [comment]
@jpetazzo I've added the ability to specify the port and removed the session id as you suggested. I think it is much better.

**Comment by marcosnils** [comment]
BTW, It'd be awesome if we add a instructions and a pre-populated url above the terminal so users can click and get redirected to port 80 right away. 

WDYT?

File: services/docker.go [hunk 1]
```diff
@@ -44,6 +45,17 @@ func CreateNetwork(name string) error {
 
 	return nil
 }
+func ConnectNetwork(containerId, networkId string) error {
+	err := c.NetworkConnect(context.Background(), networkId, containerId, &network.EndpointSettings{})
+
```
Comment (alexellis) [inline services/docker.go]: Is this change related to the reverse proxy?

Comment (marcosnils) [inline services/docker.go]: Yes because PWD container needs to be in all networks to be able to route the request.

File: api.go [hunk 2]
```diff
@@ -53,6 +53,11 @@ func main() {
 
 	r.Handle("/sessions/{sessionId}/ws/", server)
 
+	// Reverse proxy
+	proxyHandler := handlers.NewMultipleHostReverseProxy()
+	r.Host(`{node}-{port:[0-9]*}.play-with-docker.com`).Handler(proxyHandler)
+	r.Host(`{node}.play-with-docker.com`).Handler(proxyHandler)
+
```
Comment (marcosnils) [inline api.go]: > Maybe this should be an optional (configurable) component i.e.
> 
> on / off

What would be the reason to turn it off?. 





> domain name (not everyone will test / use this on "play-with-docker.com"

I agree, and I'd also like to test this locally. Can we remove the domain from the URL and assert by regex matching @xetorthio ?

**Comment by xetorthio** [comment]
We could take out the domain. But to test locally it won't help you. You
still need to able to route whatever domain to 127.0.0.1

It test this locally you'll have to set the Host header of the http request
by hand.

File: api.go (hunk 2 above)
Comment (akalipetis) [inline api.go]: We could have the base hostname in an environment variable? That would be the most straight-forward case IMO.

**Comment by akalipetis** [comment]
> It test this locally you'll have to set the Host header of the http request
by hand.

Adding hosts to `/etc/hosts` should help too I think @xetorthio 

You could also use something like [typicode/hotel](https://github.com/typicode/hotel), while I dislike such solutions 😛

**Comment by alexellis** [comment]
I think it would make more sense to switch over to flags than additional environmental variables. I introduced them initially as a quick and dirty way to enable me to contribute to the project.

Flags give help messages and are far less surprising/cryptic to the end user.

I guess when I look at this code/project I'm thinking of how it can be used not only on the official domain - but as a service by other people. That's why I've suggested being able to turn on/off various features such as the captcha and potential Docker HUB API authorisation.

**Comment by marcosnils** [comment]
> I guess when I look at this code/project I'm thinking of how it can be used not only on the official domain - but as a service by other people. That's why I've suggested being able to turn on/off various features such as the captcha and potential Docker HUB API authorisation.

Makes sense.

Regarding flags VS env variables I don't have any preference, @alexellis has a point that flags provide better context and are usually more straight forward.

**Comment by xetorthio** [comment]
Changed it so it doesn't depend on any specific TLD. Can we merge now?

**Comment by marcosnils** [comment]
LGTM. Are we going to add the sample url in the instance card in this PR or shall we do a different one?

**Comment by xetorthio** [comment]
On a different PR
